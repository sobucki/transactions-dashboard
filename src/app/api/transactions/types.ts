export type Transaction = {
  date: number;
  amount: string;
  transaction_type: TransactionType;
  currency: string;
  account: string;
  industry: string;
  state: string;
};

export type TransactionType = "deposit" | "withdraw";

export type FilterOptions = {
  startDate?: string;
  endDate?: string;
  minAmount?: string;
  maxAmount?: string;
  transactionType?: TransactionType;
  currency?: string;
  account?: string;
  industry?: string;
  state?: string;
};

export type SortOptions = {
  sortBy: keyof Transaction;
  sortOrder: "asc" | "desc";
};
