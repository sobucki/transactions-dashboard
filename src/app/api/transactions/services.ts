import { endOfDay, parseISO, startOfDay } from "date-fns";
import { FilterOptions, Transaction } from "./types";

export function filterTransactions(
  data: Transaction[],
  filter: FilterOptions
  // sortBy: SortOptions
): Transaction[] {
  let filteredData = data;

  if (filter.startDate) {
    const startDate = parseISO(filter.startDate);
    const startDateOfDay = startOfDay(startDate);

    filteredData = filteredData.filter(
      (transaction) =>
        filter.startDate && transaction.date >= startDateOfDay.getTime()
    );
  }

  if (filter.endDate) {
    const endDate = parseISO(filter.endDate);
    const endDateOfDay = endOfDay(endDate);

    filteredData = filteredData.filter(
      (transaction) =>
        filter.startDate && transaction.date <= endDateOfDay.getTime()
    );
  }

  if (filter.transactionType) {
    filteredData = filteredData.filter(
      (transaction) => transaction.transaction_type === filter.transactionType
    );
  }

  return filteredData;
}
