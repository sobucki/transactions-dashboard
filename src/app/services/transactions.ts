import { TransactionAmountSum } from "../api/transactions/types";

export async function fetchAndCalculateBalance(): Promise<string> {
  const apiUrl = "/api/sum/transaction_type";

  try {
    const response = await fetch(apiUrl, { cache: "default" });
    if (!response.ok) {
      throw new Error(`Failed to fetch: ${response.statusText}`);
    }

    const transactions: TransactionAmountSum[] = await response.json();

    const totalDeposit = transactions
      .filter((t) => t.transaction_type === "deposit")
      .reduce((sum, t) => sum + t.amount_sum, 0);

    const totalWithdraw = transactions
      .filter((t) => t.transaction_type === "withdraw")
      .reduce((sum, t) => sum + t.amount_sum, 0);

    const balance = totalDeposit - totalWithdraw;

    return (balance / 100).toLocaleString("pt-BR", {
      style: "currency",
      currency: "BRL",
    });
  } catch (error) {
    console.error("Error fetching transactions:", error);
    throw new Error("Failed to fetch transactions");
  }
}

export async function fetchAndCalculateRevenue(): Promise<string> {
  const apiUrl = "/api/sum/transaction_type?transactionType=deposit";

  try {
    const response = await fetch(apiUrl, { cache: "default" });
    if (!response.ok) {
      throw new Error(`Failed to fetch: ${response.statusText}`);
    }

    const transactions: TransactionAmountSum[] = await response.json();

    const totalDeposit = transactions.reduce((sum, t) => sum + t.amount_sum, 0);

    return (totalDeposit / 100).toLocaleString("pt-BR", {
      style: "currency",
      currency: "BRL",
    });
  } catch (error) {
    console.error("Error fetching transactions:", error);
    throw new Error("Failed to fetch transactions");
  }
}

export async function fetchAndCalculateDrawn(): Promise<string> {
  const apiUrl = "/api/sum/transaction_type?transactionType=withdraw";

  try {
    const response = await fetch(apiUrl, { cache: "default" });
    if (!response.ok) {
      throw new Error(`Failed to fetch: ${response.statusText}`);
    }

    const transactions: TransactionAmountSum[] = await response.json();

    const totalDrawn = transactions.reduce((sum, t) => sum + t.amount_sum, 0);

    return (totalDrawn / 100).toLocaleString("pt-BR", {
      style: "currency",
      currency: "BRL",
    });
  } catch (error) {
    console.error("Error fetching transactions:", error);
    throw new Error("Failed to fetch transactions");
  }
}
