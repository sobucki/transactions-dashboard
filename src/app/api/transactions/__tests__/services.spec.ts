import { filterTransactions } from "../services";
import { Transaction } from "../types";
import transactionsData from "./mock.json";

describe("services tests", () => {
  describe("filters", () => {
    describe("dates", () => {
      it("should filter transactions by start and end date", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          startDate: "2024-01-01",
          endDate: "2024-10-17",
        });

        expect(result).toEqual([
          {
            account: "Mondelez International",
            amount: "3462",
            currency: "brl",
            date: 1729195017000,
            industry: "Food Consumer Products",
            state: "IL",
            transaction_type: "deposit",
          },
        ]);

        const result2 = filterTransactions(transactionsData as Transaction[], {
          startDate: "2025-01-01",
          endDate: "2024-01-01",
        });

        expect(result2).toEqual([]);
      });
    });

    describe("type", () => {
      it("should filter transactions by transaction type deposit", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          transactionType: "deposit",
        });

        expect(result).toEqual([
          {
            date: 1729195017000,
            amount: "3462",
            transaction_type: "deposit",
            currency: "brl",
            account: "Mondelez International",
            industry: "Food Consumer Products",
            state: "IL",
          },
          {
            date: 1637751477490,
            amount: "3462",
            transaction_type: "deposit",
            currency: "brl",
            account: "Mondelez International",
            industry: "Food Consumer Products",
            state: "IL",
          },

          {
            date: 1637751477490,
            amount: "2000",
            transaction_type: "deposit",
            currency: "eur",
            account: "Tesla Inc",
            industry: "Automotive",
            state: "CA",
          },

          {
            date: 1609459200000,
            amount: "3000",
            transaction_type: "deposit",
            currency: "usd",
            account: "Google",
            industry: "Technology",
            state: "CA",
          },

          {
            date: 1637751477490,
            amount: "1000",
            transaction_type: "deposit",
            currency: "usd",
            account: "Facebook",
            industry: "Social Media",
            state: "CA",
          },

          {
            date: 1609459200000,
            amount: "1200",
            transaction_type: "deposit",
            currency: "usd",
            account: "Twitter",
            industry: "Social Media",
            state: "CA",
          },
        ]);
      });

      it("should filter transactions by transaction type withdraw", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          transactionType: "withdraw",
        });

        expect(result).toEqual([
          {
            account: "Apple Inc",
            amount: "1500",
            currency: "usd",
            date: 1637751477490,
            industry: "Technology",
            state: "CA",
            transaction_type: "withdraw",
          },
          {
            account: "Amazon",
            amount: "500",
            currency: "usd",
            date: 1609459200000,
            industry: "E-commerce",
            state: "WA",
            transaction_type: "withdraw",
          },
          {
            account: "Microsoft",
            amount: "2500",
            currency: "usd",
            date: 1609459200000,
            industry: "Technology",
            state: "WA",
            transaction_type: "withdraw",
          },
          {
            account: "Netflix",
            amount: "750",
            currency: "usd",
            date: 1637751477490,
            industry: "Entertainment",
            state: "CA",
            transaction_type: "withdraw",
          },
          {
            account: "Uber",
            amount: "1800",
            currency: "usd",
            date: 1609459200000,
            industry: "Transportation",
            state: "CA",
            transaction_type: "withdraw",
          },
        ]);
      });
    });

    describe("currency", () => {
      it("should filter transactions by currency brl", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          currency: "brl",
          transactionType: "deposit",
        });

        expect(result).toEqual([
          {
            account: "Mondelez International",
            amount: "3462",
            currency: "brl",
            date: 1729195017000,
            industry: "Food Consumer Products",
            state: "IL",
            transaction_type: "deposit",
          },
          {
            account: "Mondelez International",
            amount: "3462",
            currency: "brl",
            date: 1637751477490,
            industry: "Food Consumer Products",
            state: "IL",
            transaction_type: "deposit",
          },
        ]);
      });

      it("should filter transactions by currency euro", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          currency: "eur",
        });

        expect(result).toEqual([
          {
            account: "Tesla Inc",
            amount: "2000",
            currency: "eur",
            date: 1637751477490,
            industry: "Automotive",
            state: "CA",
            transaction_type: "deposit",
          },
        ]);
      });
    });

    describe("account", () => {
      it("should filter transactions by account name", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          account: "Apple Inc",
        });

        expect(result).toEqual([
          {
            account: "Apple Inc",
            amount: "1500",
            currency: "usd",
            date: 1637751477490,
            industry: "Technology",
            state: "CA",
            transaction_type: "withdraw",
          },
        ]);
      });
    });

    describe("industry", () => {
      it("should filter transactions by industry name", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          industry: "Technology",
        });

        expect(result).toEqual([
          {
            date: 1637751477490,
            amount: "1500",
            transaction_type: "withdraw",
            currency: "usd",
            account: "Apple Inc",
            industry: "Technology",
            state: "CA",
          },
          {
            account: "Google",
            amount: "3000",
            currency: "usd",
            date: 1609459200000,
            industry: "Technology",
            state: "CA",
            transaction_type: "deposit",
          },
          {
            account: "Microsoft",
            amount: "2500",
            currency: "usd",
            date: 1609459200000,
            industry: "Technology",
            state: "WA",
            transaction_type: "withdraw",
          },
        ]);
      });
    });

    describe("state", () => {
      it("should filter transactions by state name", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          state: "IL",
        });

        expect(result).toEqual([
          {
            account: "Mondelez International",
            amount: "3462",
            currency: "brl",
            date: 1729195017000,
            industry: "Food Consumer Products",
            state: "IL",
            transaction_type: "deposit",
          },
          {
            account: "Mondelez International",
            amount: "3462",
            currency: "brl",
            date: 1637751477490,
            industry: "Food Consumer Products",
            state: "IL",
            transaction_type: "deposit",
          },
        ]);
      });
    });

    describe("amount", () => {
      it("should filter transactions by max amount", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          maxAmount: "1000",
        });

        expect(result).toEqual([
          {
            account: "Amazon",
            amount: "500",
            currency: "usd",
            date: 1609459200000,
            industry: "E-commerce",
            state: "WA",
            transaction_type: "withdraw",
          },
          {
            account: "Facebook",
            amount: "1000",
            currency: "usd",
            date: 1637751477490,
            industry: "Social Media",
            state: "CA",
            transaction_type: "deposit",
          },
          {
            account: "Netflix",
            amount: "750",
            currency: "usd",
            date: 1637751477490,
            industry: "Entertainment",
            state: "CA",
            transaction_type: "withdraw",
          },
        ]);
      });

      it("should filter transactions by min amount", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          minAmount: "3000",
        });

        expect(result).toEqual([
          {
            account: "Mondelez International",
            amount: "3462",
            currency: "brl",
            date: 1729195017000,
            industry: "Food Consumer Products",
            state: "IL",
            transaction_type: "deposit",
          },
          {
            account: "Mondelez International",
            amount: "3462",
            currency: "brl",
            date: 1637751477490,
            industry: "Food Consumer Products",
            state: "IL",
            transaction_type: "deposit",
          },
          {
            account: "Google",
            amount: "3000",
            currency: "usd",
            date: 1609459200000,
            industry: "Technology",
            state: "CA",
            transaction_type: "deposit",
          },
        ]);
      });

      it("should filter transactions by min and max amount", () => {
        const result = filterTransactions(transactionsData as Transaction[], {
          minAmount: "1000",
          maxAmount: "1500",
        });

        expect(result).toEqual([
          {
            account: "Apple Inc",
            amount: "1500",
            currency: "usd",
            date: 1637751477490,
            industry: "Technology",
            state: "CA",
            transaction_type: "withdraw",
          },
          {
            account: "Facebook",
            amount: "1000",
            currency: "usd",
            date: 1637751477490,
            industry: "Social Media",
            state: "CA",
            transaction_type: "deposit",
          },
          {
            account: "Twitter",
            amount: "1200",
            currency: "usd",
            date: 1609459200000,
            industry: "Social Media",
            state: "CA",
            transaction_type: "deposit",
          },
        ]);
      });
    });
  });
});
