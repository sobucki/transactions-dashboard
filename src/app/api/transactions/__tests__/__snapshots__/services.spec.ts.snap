// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services tests > sort > should sort transactions by amount in ascending order 1`] = `
[
  {
    "account": "Apple Inc",
    "amount": "1500",
    "currency": "usd",
    "date": 1637751477490,
    "industry": "Technology",
    "state": "CA",
    "transaction_type": "withdraw",
  },
  {
    "account": "Tesla Inc",
    "amount": "2000",
    "currency": "eur",
    "date": 1637751477490,
    "industry": "Automotive",
    "state": "CA",
    "transaction_type": "deposit",
  },
  {
    "account": "Microsoft",
    "amount": "2500",
    "currency": "usd",
    "date": 1609459200000,
    "industry": "Technology",
    "state": "WA",
    "transaction_type": "withdraw",
  },
  {
    "account": "Google",
    "amount": "3000",
    "currency": "usd",
    "date": 1609459200000,
    "industry": "Technology",
    "state": "CA",
    "transaction_type": "deposit",
  },
]
`;
